---
title: 'Intro to phylogenetics: practical'
subtitle: 'AIMS-Imperial: modern statistics for global health'
author: "Alexandra Blenkinsop"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Objectives

This practical will show you how to run a phylogenetic pipeline from consensus sequences including creating an alignment, specifying a suitable outgroup for rooting, inferring a phylogeny using the distance-based neighbour-joining method, resolving alignment issues and using bootstrapping to quantify uncertainty.

We will be using the \texttt{ape} package for this practical, which has several tools for manipulating DNA sequence data in R and carrying out phylogenetic inference. The first step is to make sure you have this installed, as well as the other packages we need, using \texttt{install.packages()}.

# Load packages
```{r, include=TRUE, eval=TRUE, message=FALSE, echo=TRUE, warning=FALSE, tidy=TRUE}
require(data.table)
require(knitr)
# phylo packages
require(ape)
require(ade4)
# packages for plotting
require(ggplot2)
require(ggtree)
require(ggsci)
require(viridis)

```

## Data

Download the consensus sequences for this practical from \url{}, which are a subset of partial polyamerase (\textit{pol}) sequences from HIV positive individuals from the Democratic Republic of Congo from this study by [Faria et al, 2019]{https://doi.org/10.1371/journal.ppat.1007976}, tracing the evolutionary dynamics of major HIV-1 subtypes across Central and Eastern Africa. The data were obtained from the [Los Alamos HIV sequence database]{https://www.hiv.lanl.gov/components/sequence/HIV/search/search.html}, by searching for the GenBank Accession numbers listed in the manuscript and selecting a subset of subtypes for the purpose of this practical.

To view the alignemt, you can download an alignment viewer such as [AliView]{https://ormbunkar.se/aliview/}, or use the [IGV web interface]{https://igv.org/app/} or the [National Center for Biotechnology Information Multiple Sequence Alignment Viewer interface]{https://www.ncbi.nlm.nih.gov/projects/msaviewer/}. Here are the first ~360 positions of the alignment:

```{r, include=TRUE, fig.height=8, fig.width=10, eval=TRUE, message=FALSE, echo=TRUE, warning=FALSE, tidy=TRUE}

data.dir <- '/Users/alexb/Documents/GitHub/Teaching/phylogenetics'
out.dir <- '/Users/alexb/Documents/GitHub/Teaching/phylogenetics'

knitr::include_graphics(file.path(data.dir,'faria_2019_alignment.png'))

```


The taxa names are named uniformly in the Los Alamos HIV sequence database, with the format <Subtype.Isocode.Year.Name.Accession>.

## Read in the alignment

Your first task is to load the multiple sequence alignment and have a look at its structure.

Read in the .fasta file. Note the format of FASTA - chevron denotes the taxon name. Gaps '-' are indel events (insertions/deletions). 'N' represents unknown characters.

How many sequences do we have and what length are they?


```{r, out.width='100%',include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# read data
filename <- file.path(data.dir,'faria_2019.fasta')

dna <- read.dna(file = filename, format = "fasta")
dna
class(dna)
# look at positions 1:10 in the alignment of the first 5 taxa
as.character(dna)[1:5, 1:10]

```

## Generate the genetic distance matrix

Generate the distance matrix using the \texttt{ape} package, assuming the Tamura and Nei model for the evolutionary rate, which has two parameters. Plot the matrix.

Note that entries $i=j$ have a genetic distance of zero.

There are two types of HIV, HIV-1 and HIV-2, which can further be classified by subtype. All the sequences in this study are from individuals with HIV-1, which is the most prevalent type of HIV globally. The taxa labels indicate subtype. Subtypes are an indication of virus similarity. Which sequences have the largest distances?

```{r, out.width='100%', include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# calculate genetic distances using Tamura and Nei model (diff rates of transitions and transversions & bween site variation of substitution rate)
D <- dist.dna(dna, model = "TN93")

# plot distance matrix
temp <- as.data.frame(as.matrix(D))
temp <- t(as.matrix(D))
temp <- temp[, ncol(temp):1]
par(mar = c(1, 5, 5, 1))
nseqs <- dim(dna)[1]

# melt the matrix to a data.table 
tmp <- data.table(reshape2::melt(temp))
# ensure the ordering of taxa are the same for plotting the matrix
tmp[, Var2:= factor(Var2,levels=levels(Var1))]

ggplot(tmp) + geom_tile(aes(x=Var1,y=Var2,fill=value)) +
  scale_fill_viridis(option="magma",direction=-1) + 
  labs(x='',y='',fill='Genetic distances\n(Tamura and Nei model)') +
  theme_bw(base_size=6) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
 
```


## Build the tree

We can use the ape package in R, which has a few algorithms available. Build a tree using the neighbour-joining algorithm.

What do you notice? Do any of the tree features suggest something has gone wrong with the alignment or phylogenetic inference?


```{r, out.width='100%', out.height='100%', fig.width=6, fig.height=6, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}
 
select <- data.table(taxa=labels(dna))
select[, ST:=gsub('([A-Za-z0-9_]+).*','\\1',taxa)]
mypal <- pal_npg('nrc')(4)
select[, ST:= factor(ST)]

colmap <- data.table(ST=c('G','02_AG','A','Ref'),
                     color=mypal)
select <- merge(select,colmap,by='ST')
cols <- with(select, as.character(color[match(labels(dna), taxa)]))

# build tree
tre <- nj(D)
class(tre)
tre
plot(tre, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Tree using neighbour-joining algorithm")

#ggtree(nhx) + geom_tiplab(aes(color = label %in% to_drop)) +
#  scale_color_manual(values=c("black", "red")) + xlim(0, 0.8)
summary(tre)

```


## Correct the alignment

See if you can find the reason for the long branch by taking a look at the alignment.

Open the .fasta file with an alignment viewer and see if you can find and correct the alignment error. Hint: the sequences in the alignment must all be the same length.

Save it with a different name.

Read in the new .fasta file and repeat the steps above. Now the distance matrix no longer shows large differences between all sequences and the misligned sequence. The long branch has also gone from the tree.

What else do you notice? Hint: the tips of the tree have been coloured according to their HIV-1 subtype. 

 
```{r, out.width='100%', out.height='100%', include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}
filename <- file.path(data.dir,'faria_2019_fixedalignment.fasta')

dna <- read.dna(file = filename, format = "fasta")

D <- dist.dna(dna, model = "TN93")

# plot distance matrix
temp <- as.data.frame(as.matrix(D))
temp <- t(as.matrix(D))
temp <- temp[, ncol(temp):1]
par(mar = c(1, 5, 5, 1))
nseqs <- dim(dna)[1]

# melt the matrix to a data.table 
tmp <- data.table(reshape2::melt(temp))
# ensure the ordering of taxa are the same for plotting the matrix
tmp[, Var2:= factor(Var2,levels=levels(Var1))]

ggplot(tmp) + geom_tile(aes(x=Var1,y=Var2,fill=value)) +
  scale_fill_viridis(option="magma",direction=-1) + 
  labs(x='',y='',fill='Genetic distances\n(Tamura and Nei model)') +
  theme_bw(base_size=6) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

select <- data.table(taxa=labels(dna))
select[, ST:=gsub('([A-Za-z0-9_]+).*','\\1',taxa)]
mypal <- pal_npg('nrc')(4)
select[, ST:= factor(ST)]

colmap <- data.table(ST=c('G','02_AG','A','Ref'),
                     color=mypal)
select <- merge(select,colmap,by='ST')
cols <- with(select, as.character(color[match(labels(dna), taxa)]))

# build tree
tre <- nj(D)
class(tre)
tre
plot(tre, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Tree using neighbour-joining algorithm")

#ggtree(nhx) + geom_tiplab(aes(color = label %in% to_drop)) +
#  scale_color_manual(values=c("black", "red")) + xlim(0, 0.8)
summary(tre)

```


We notice that there is a recombinant of subtypes A and G. The algorithm doesn't know where to put it in the tree, and we have a clade of both subtype A and subtype G viruses. What happens if we remove it?

Hint: the taxa labels for the alignment are stored in \texttt{labels(dna)}.

# Remove the recombinant
```{r, out.width='100%',fig.width=8, fig.height=10, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

# find recombinant taxa label which isn't subtype G or A
to_remove <- labels(dna)[!gsub('([A-Za-z0-9_]+).*','\\1',labels(dna)) %in% c('G','A','Ref')]

# remove recombinant
dna <- dna[!labels(dna)==to_remove,]

select <- data.table(taxa=labels(dna))
select[, ST:=gsub('([A-Za-z0-9_]+).*','\\1',taxa)]
mypal <- pal_npg('nrc')(4)
select[, ST:= factor(ST)]

D <- dist.dna(dna, model = "TN93")

select[, ST:=gsub('([A-Za-z0-9_]+).*','\\1',taxa)]
mypal <- pal_npg('nrc')(4)
select[, ST:= factor(ST)]

colmap <- data.table(ST=c('G','02_AG','A','Ref'),
                     color=mypal)
select <- merge(select,colmap,by='ST')
cols <- with(select, as.character(color[match(labels(dna), taxa)]))

tre <- nj(D)
plot(tre, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Unrooted tree using neighbour-joining algorithm")


```


The reference sequence (subtype C) is part of the G clade, and the subtype A sequences are all in subclades, which is not what we expect given the subtype classifications. We can tell the algorithm where the root should be by specifying the outgroup.

We can root the tree using the command \texttt{root()}, and we can drop the root tip from the phylogeny as it isn't part of our study, using the \texttt{drop.tip()} function in \texttt{ape}.

## Root the tree
```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

tre2 <- root(tre, outgroup = 'Ref.C.ZA.04.04ZASK146.AY772699',resolve.root=TRUE)
plot(tre2, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("Rooted tree using neighbour-joining algorithm")


```

Now by rooting the tree, the reference sequence has moved to be the ancestor of all the taxa. We now also see that all the subtype G taxa are in one clade, and all the subtype A taxa are in another.


[Questions about interpretation]

who is related?
who shares common ancestor []?

## Bootstrapping

Now let's bootstrap resample each site of the alignment with replacement, carry out phylogenetic inference as before, and plot the tree with internal nodes labelled with the number of boostrapped trees 

```{r, out.width='100%',fig.width=8, fig.height=8, include=TRUE, echo=TRUE, eval=TRUE, fig.align="center"}

#D <- dist.dna(dna, model = "TN93")
#tre <- nj(D)
#tre2 <- root(tre, outgroup = 'Ref.C.ZA.04.04ZASK146.AY772699',resolve.root=TRUE)

myBoots <- boot.phylo(tre2, dna, function(e) nj(dist.dna(e, model = "TN93")),B=100)
#myBoots


plot(tre2, cex = 0.6, tip.color = cols)
add.scale.bar(x=0,y=0)
title("NJ tree + bootstrap values")
#tiplabels(frame = "none", pch = 20, col = transp(num2col(annot$year,
#+ col.pal = myPal), 0.7), cex = 3, fg = "transparent")
#axisPhylo()
#temp <- pretty(1993:2008, 5)
#legend("topright", fill = transp(num2col(temp, col.pal = myPal),
#+ 0.7), leg = temp, ncol = 2)
nodelabels(myBoots, cex = 0.5)

#d <- dist.dna(woodmouse)
#tr <- nj(d)
#bp <- boot.phylo(tr, woodmouse, function(xx) nj(dist.dna(xx)))
#tree <- apeBoot(tr, bp)
#ggtree(tree) + geom_label(aes(label=bootstrap)) + geom_tiplab()

```

## Re-build the phylogeny using a Bayesian approach

stan

